app/src/main/java/cm/aptoide/pt/networking/IdsRepository.java

```diff
@@ -43,95 +44,92 @@ public IdsRepository(SharedPreferences sharedPreferences, Context context, Strin
-    }
-
-    if (AdNetworkUtils.isGooglePlayServicesAvailable(context)) {
-      try {
-        googleAdvertisingId = AdvertisingIdClient.getAdvertisingIdInfo(context)
-            .getId();
-      } catch (Exception e) {
-        CrashReport.getInstance()
-            .log(e);
-      }
-    }
-
-    sharedPreferences.edit()
-        .putString(GOOGLE_ADVERTISING_ID_CLIENT, googleAdvertisingId)
-        .apply();
-    sharedPreferences.edit()
-        .putBoolean(GOOGLE_ADVERTISING_ID_CLIENT_SET, true)
-        .apply();
-
-    return googleAdvertisingId;
+  @WorkerThread public synchronized Single<String> getGoogleAdvertisingId() {
+
+    return Single.just(sharedPreferences.getString(GOOGLE_ADVERTISING_ID_CLIENT, null))
+        .map(id -> {
+          if (!TextUtils.isEmpty(id)) {
+            return id;
+          } else if (AptoideUtils.ThreadU.isUiThread()) {
+            throw new IllegalStateException("You cannot run this method from the main thread");
+          } else if (!AdNetworkUtils.isGooglePlayServicesAvailable(context)) {
+            return "";
+          } else {
+            try {
+              return AdvertisingIdClient.getAdvertisingIdInfo(context)
+                  .getId();
+            } catch (Exception e) {
+              throw new IllegalStateException(e);
+            }
+          }
+        })
+        .doOnSuccess(id -> {
+          if (!id.equals("")) {
+            sharedPreferences.edit()
+                .putString(GOOGLE_ADVERTISING_ID_CLIENT, id)
+                .apply();
+            sharedPreferences.edit()
+                .putBoolean(GOOGLE_ADVERTISING_ID_CLIENT_SET, true)
+                .apply();
+          }
+        })
+        .subscribeOn(Schedulers.newThread());
   }
 
-  public synchronized String getAdvertisingId() {
+  public synchronized Single<String> getAdvertisingId() {
     String advertisingId = sharedPreferences.getString(ADVERTISING_ID_CLIENT, null);
     if (!TextUtils.isEmpty(advertisingId)) {
       // if we already have an advertising id, return it
-      return advertisingId;
-    }
-
-    // else, generate using google advertising id
-    advertisingId = getGoogleAdvertisingId();
-    if (TextUtils.isEmpty(advertisingId)) {
-      // if google advertising id is not available use a random id
-      advertisingId = generateRandomAdvertisingId();
+      return Single.just(advertisingId);
     }
 
-    // save the generated advertising id for this user and return it
-    sharedPreferences.edit()
-        .putString(ADVERTISING_ID_CLIENT, advertisingId)
-        .apply();
-    return advertisingId;
+    return getGoogleAdvertisingId().map(id -> {
+      if (!TextUtils.isEmpty(id)) {
+        return id;
+      } else {
+        return "";
+      }
+    })
+        .map(id -> {
+          if (!TextUtils.isEmpty(id)) {
+            return id;
+          } else {
+            return generateRandomAdvertisingId();
+          }
+        })
```

Comment:
getGoogleAdvertisingId() returns id or empty string, this part is a bit messy, could be refactored.